src/tools/rbd_wnbd/rbd_mapping.cc

```diff
@@ -282,3 +288,129 @@ void RbdMappingDispatcher::disconnect_cbk(std::string devpath, int ret)
     mappings.erase(devpath);
   }
 }
+
+int RbdMappingDispatcher::stop(
+  bool hard_disconnect,
+  int soft_disconnect_timeout,
+  int worker_count)
+{
+  stop_requested = true;
+
+  // Although not generally recommended, soft_disconnect_timeout can be 0,
+  // which means infinite timeout.
+  ceph_assert(soft_disconnect_timeout >= 0);
+  ceph_assert(worker_count > 0);
+
+  std::atomic<int> err = 0;
+
+  boost::asio::thread_pool pool(worker_count);
+  LARGE_INTEGER start_t, counter_freq;
+  QueryPerformanceFrequency(&counter_freq);
+  QueryPerformanceCounter(&start_t);
+
+  // We're initiating the disk removal through libwnbd, which uses PNP
+  // to notify the storage stack that the disk is going to be removed
+  // and waits for pending operations to complete.
+  //
+  // Once ready, the WNBD driver notifies rbd-wnbd that the disk has been
+  // disconnected.
+  auto mapped_devpaths = get_mapped_devpaths();
+  for (auto devpath: mapped_devpaths) {
+    boost::asio::post(pool,
+      [devpath, start_t, counter_freq, soft_disconnect_timeout,
+       hard_disconnect, &err]() mutable
+      {
+        LARGE_INTEGER curr_t, elapsed_ms;
+        QueryPerformanceCounter(&curr_t);
+        elapsed_ms.QuadPart = curr_t.QuadPart - start_t.QuadPart;
+        elapsed_ms.QuadPart *= 1000;
+        elapsed_ms.QuadPart /= counter_freq.QuadPart;
+
+        int64_t time_left_ms = std::max(
+          (int64_t)0,
+          soft_disconnect_timeout * 1000 - elapsed_ms.QuadPart);
+
+        dout(1) << "Removing mapping: " << devpath
+                << ". Timeout: " << time_left_ms
+                << "ms. Hard disconnect: " << hard_disconnect
+                << dendl;
+
+        WNBD_REMOVE_OPTIONS remove_options = {0};
+        remove_options.Flags.HardRemove = hard_disconnect || !time_left_ms;
+        remove_options.Flags.HardRemoveFallback = true;
+        remove_options.SoftRemoveTimeoutMs = time_left_ms;
+        remove_options.SoftRemoveRetryIntervalMs = SOFT_REMOVE_RETRY_INTERVAL * 1000;
+
+        // This is asynchronous, it may take a few seconds for the disk to be
+        // removed. We'll perform the wait outside the loop to speed up the
+        // process.
+        int r = WnbdRemoveEx(devpath.c_str(), &remove_options);
+        if (r && r != ERROR_FILE_NOT_FOUND) {
+          err = -EINVAL;
+          derr << "Could not remove mapping: " << devpath
+               << ". Error: " << win32_strerror(r) << dendl;
+        } else {
+          dout(1) << "Successfully removed mapping: " << devpath << dendl;
+        }
+      });
+  }
+  pool.join();
+
+  if (err) {
+    derr << "Could not remove all mappings. Error: " << err << dendl;
+    return err;
+  }
+
+  // Wait for the cleanup to complete on the rbd-wnbd service side.
+  return wait_for_mappings_removal(10000);
+}
+
+std::set<std::string> RbdMappingDispatcher::get_mapped_devpaths() {
+  std::set<std::string> out;
```

Comment:
Does using a set have a significance here?  We can't get duplicate devpaths out of `mappings`, do we?

Reply:
The default, appropriate in most cases, container type is usually `std::vector`.  Using a list or a set should have a specific purpose, such as needing to delete in the front/middle, needing to sort or discard duplicates, etc.